### ISSUE TITLE
corrupted on transfer: sizes differ xxxxx vs xxxx (s3 NetApp StorageGrid -> local)

### ISSUE THREAD

**Issue by fwfichtner** [rclone/rclone#6533]
#### The associated forum post URL from `https://forum.rclone.org`

-

#### What is the problem you are having with rclone?

Similar to https://github.com/rclone/rclone/issues/5732 rclone fails to move files from s3 to local disk with the error corrupted on transfer: sizes differ xxxx vs xxxx.

I would have just re-opened the issue, but could not.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.60.0
- os/version: Microsoft Windows 10 Enterprise 21H2 (64 bit)
- os/kernel: 10.0.19044.2130 (x86_64)
- os/type: windows
- os/arch: amd64
- go/version: go1.19.2
- go/linking: static
- go/tags: cmount
```
or
```
rclone v1.60.0
- os/version: alpine 3.16.2 (64 bit)
- os/kernel: 3.10.0-1160.71.1.el7.x86_64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.19.2
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Microsoft Windows 10 Enterprise 21H2 (64 bit) or alpine 3.16.2 (64 bit).

#### Which cloud storage system are you using? (e.g. Google Drive)

S3-compliant storage.

```
[src]
type = s3
provider = other
env_auth = false
access_key_id = xxx
secret_access_key = xxx
endpoint = xx.xxxx.xxx:10443
```

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone copy remote:tmp\test.txt /tmp`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
rclone copy netapp_s3:tmp/test.txt C:\Users\username\Downloads -vv --dump bodies
2022/11/02 11:13:23 DEBUG : rclone: Version "v1.60.0" starting with parameters ["rclone" "copy" "netapp_s3:tmp/test.txt" "C:\\Users\\username\\Downloads" "-vv" "--dump" "bodies"]
2022/11/02 11:13:23 DEBUG : Creating backend with remote "netapp_s3:tmp/test.txt"
2022/11/02 11:13:23 DEBUG : Using config file from "C:\\Users\\username\\.config\\rclone\\rclone.conf"
2022/11/02 11:13:23 DEBUG : You have specified to dump information. Please be noted that the Accept-Encoding as shown may not be correct in the request and the response may not show Content-Encoding if the go standard libraries auto gzip encoding was in effect. In this case the body of the request will be gunzipped before showing it.
2022/11/02 11:13:23 DEBUG : You have specified to dump information. Please be noted that the Accept-Encoding as shown may not be correct in the request and the response may not show Content-Encoding if the go standard libraries auto gzip encoding was in effect. In this case the body of the request will be gunzipped before showing it.
2022/11/02 11:13:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/11/02 11:13:23 DEBUG : HTTP REQUEST (req 0xc00085dd00)
2022/11/02 11:13:23 DEBUG : HEAD /tmp/test.txt HTTP/1.1
Host: example.url
User-Agent: rclone/v1.60.0
Authorization: XXXX
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20221102T101323Z

2022/11/02 11:13:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/11/02 11:13:23 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/11/02 11:13:23 DEBUG : HTTP RESPONSE (req 0xc00085dd00)
2022/11/02 11:13:23 DEBUG : HTTP/1.1 200 OK
Content-Length: 11
Accept-Ranges: bytes
Connection: keep-alive
Content-Type: text/plain; charset=utf-8
Date: Wed, 02 Nov 2022 10:13:23 GMT
Etag: "5eb63bbbe01eeed093cb22bb8f5acdc3"
Last-Modified: Wed, 02 Nov 2022 09:36:21 GMT
Server: StorageGRID/11.5.0.6
X-Amz-Expiration: expiry-date="Thu, 03 Nov 2022 09:36:21 GMT", rule-id="59fb63d7-4a45-4cad-a5f3-d5511187c3bc"
X-Amz-Id-2: 12725755
X-Amz-Meta-Mtime: 1667381721.3128405
X-Amz-Request-Id: 1667382149450076
X-Ntap-Sg-Trace-Id: 7c9988d56c41cf81

2022/11/02 11:13:23 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/11/02 11:13:23 DEBUG : fs cache: adding new entry for parent of "netapp_s3:tmp/test.txt", "netapp_s3:tmp"
2022/11/02 11:13:23 DEBUG : Creating backend with remote "C:\\Users\\username\\Downloads"
2022/11/02 11:13:23 DEBUG : fs cache: renaming cache item "C:\\Users\\username\\Downloads" to be canonical "//?/C:/Users/username/Downloads"
2022/11/02 11:13:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/11/02 11:13:23 DEBUG : HTTP REQUEST (req 0xc0009a8300)
2022/11/02 11:13:23 DEBUG : HEAD /tmp/test.txt HTTP/1.1
Host: example.url
User-Agent: rclone/v1.60.0
Authorization: XXXX
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20221102T101323Z

2022/11/02 11:13:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/11/02 11:13:23 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/11/02 11:13:23 DEBUG : HTTP RESPONSE (req 0xc0009a8300)
2022/11/02 11:13:23 DEBUG : HTTP/1.1 200 OK
Content-Length: 11
Accept-Ranges: bytes
Connection: keep-alive
Content-Type: text/plain; charset=utf-8
Date: Wed, 02 Nov 2022 10:13:23 GMT
Etag: "5eb63bbbe01eeed093cb22bb8f5acdc3"
Last-Modified: Wed, 02 Nov 2022 09:36:21 GMT
Server: StorageGRID/11.5.0.6
X-Amz-Expiration: expiry-date="Thu, 03 Nov 2022 09:36:21 GMT", rule-id="59fb63d7-4a45-4cad-a5f3-d5511187c3bc"
X-Amz-Id-2: 12293467
X-Amz-Meta-Mtime: 1667381721.3128405
X-Amz-Request-Id: 1667382149050734
X-Ntap-Sg-Trace-Id: dac2cfa4cd1533f7

2022/11/02 11:13:23 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/11/02 11:13:23 DEBUG : test.txt: Need to transfer - File not found at Destination
2022/11/02 11:13:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/11/02 11:13:23 DEBUG : HTTP REQUEST (req 0xc00051a200)
2022/11/02 11:13:23 DEBUG : GET /tmp/test.txt HTTP/1.1
Host: example.url
User-Agent: rclone/v1.60.0
Accept-Encoding: gzip
Authorization: XXXX
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20221102T101323Z

2022/11/02 11:13:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/11/02 11:13:23 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/11/02 11:13:23 DEBUG : HTTP RESPONSE (req 0xc00051a200)
2022/11/02 11:13:23 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Accept-Ranges: bytes
Connection: keep-alive
Content-Encoding: gzip
Content-Type: text/plain; charset=utf-8
Date: Wed, 02 Nov 2022 10:13:23 GMT
Etag: "5eb63bbbe01eeed093cb22bb8f5acdc3"
Last-Modified: Wed, 02 Nov 2022 09:36:21 GMT
Server: StorageGRID/11.5.0.6
X-Amz-Expiration: expiry-date="Thu, 03 Nov 2022 09:36:21 GMT", rule-id="59fb63d7-4a45-4cad-a5f3-d5511187c3bc"
X-Amz-Id-2: 12725755
X-Amz-Meta-Mtime: 1667381721.3128405
X-Amz-Request-Id: 1667382149450076
X-Ntap-Sg-Trace-Id: fc121a3ff62d0358

1f
♂     ♦♥�H���W(�/�I☺ �◄J
0

2022/11/02 11:13:23 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2022/11/02 11:13:23 NOTICE: test.txt: Not decompressing 'Content-Encoding: gzip' compressed file. Use --s3-decompress to override
2022/11/02 11:13:23 ERROR : test.txt: corrupted on transfer: sizes differ 11 vs 31
2022/11/02 11:13:23 INFO  : test.txt: Removing failed copy
2022/11/02 11:13:23 ERROR : Attempt 1/3 failed with 1 errors and: corrupted on transfer: sizes differ 11 vs 31
2022/11/02 11:13:23 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2022/11/02 11:13:23 DEBUG : HTTP REQUEST (req 0xc00051a700)
2022/11/02 11:13:23 DEBUG : HEAD /tmp/test.txt HTTP/1.1
Host: example.url
User-Agent: rclone/v1.60.0
Authorization: XXXX
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20221102T101323Z
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
A new issue is best - thanks.

Which provider are you using? StorageGRID?

I think this is the same issue as we've been discussing here: https://forum.rclone.org/t/s3-error-corrupted-on-transfer-sizes-differ-nnn-vs-mmm/33694/

I've tried to fix this with a provider quirk - can you give this a go?

[v1.61.0-beta.6506.b615cbeaa.fix-r2-gzip](https://beta.rclone.org/branch/fix-r2-gzip/v1.61.0-beta.6506.b615cbeaa.fix-r2-gzip/) on branch [fix-r2-gzip](https://github.com/rclone/rclone/tree/fix-r2-gzip) (uploaded in 15-30 mins)

This should be set automatically for `provider = Cloudflare` but you can set it manually with --s3-might-gzip=true which might help you.

**Comment by fwfichtner** [comment]
Thanks!

I am using StorageGRID and it works using your fix when I manually set --s3-might-gzip=true.

**Comment by ncw** [comment]
Great - thanks for testing.

Hmm, I wonder if I should set the default (Other) provider to `might_gzip = true`.

Or maybe make `might_gzip=true` for all providers unless proved otherwise.

The patch looks for `transfer-encoding: chunked` and `content-encoding: gzip to engage the fix. This is reasonably sensitive I think as I don't think AWS ever uses chunked encoding when downloading objects so by definition its out of spec.

Any thoughts?

**Comment by fwfichtner** [comment]
I would suggest doing it for all providers if you don't see any negative impact for now.

**Comment by ncw** [comment]
I agree with you. I've changed the default to be `might_gzip = true` for all providers except AWS. This should mean you can use this binary with no extra flags.

[v1.61.0-beta.6506.a6295901f.fix-r2-gzip](https://beta.rclone.org/branch/fix-r2-gzip/v1.61.0-beta.6506.a6295901f.fix-r2-gzip/) on branch [fix-r2-gzip](https://github.com/rclone/rclone/tree/fix-r2-gzip) (uploaded in 15-30 mins)

Can you see if this works for you?

I'll probably put this patch in 1.60.1 as quite a lot of people have been having trouble with this.

**Comment by fwfichtner** [comment]
Works for me!

**Comment by ncw** [comment]
Thanks for testing. I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.61 - I've also put it in the queue for v1.60.1

**Comment by fwfichtner** [comment]
Thank you very much for the fast fix!
